### DISCUSSION TITLE
act runner

### DISCUSSION THREAD

**Discussion by Mohammad-Alqadi** [nektos/act#1316] [Q&A] [answered]
hey there
I'm not sure if i have missed something, but i downloaded act and start working on it and everything went good
But i'm not able to see the container the act works on.
I have deleted all the containers and the images i have locally just to force act to rebuild everything over again but it didn't mind it at all and worked again so fast like nothing happened.
also i use one step on the workflow to write some content into file, i was trying to find where this file is but i couldn't 

Just to make sure that i'm getting this right
Where the (runners/docker containers) of act are?
does act run on some global containers? (is my repo safe?)

**Comment by Antvirf** [comment, marked as answer]
## On observing (running) ```act``` containers
If you start a flow with ```act``` (and assuming it is still running), you should be able to see the container with ```docker ps```. Once it's finished, it will be deleted though. (If on linux you may need to prefix ```docker``` commands with sudo depending on your setup.)


For example, if I just started a flow, I can see it with:
```
$ docker ps
CONTAINER ID   IMAGE                           COMMAND                  CREATED         STATUS        PORTS     NAMES
fe9ece598cdf   catthehacker/ubuntu:act-20.04   "/usr/bin/tail -f /d…"   2 seconds ago   Up 1 second             name-of-the-thing-you-are-working-on
```

## On where the runners are
They should be visible with ```docker image ls```, just like any other image.
```
$ docker image ls
REPOSITORY                                                TAG                                                                          IMAGE ID       CREATED         SIZE
catthehacker/ubuntu                                       act-20.04                                                                    cf0a225bbc9e   3 days ago      1.14GB
catthehacker/ubuntu                                       act-latest                                                                   a6a1365ccd00   10 days ago     1.14GB
```

If you remove an image, then it will need to pull it again when it runs. For example, I have a workflow with ubuntu-18.04 and I deleted it - when I try to run that flow:
```
$ act   
[Test dual-trigger workflow/test_dual_trigger] 🚀  Start image=catthehacker/ubuntu:act-18.04
[Test dual-trigger workflow/test_dual_trigger]   🐳  docker pull image=catthehacker/ubuntu:act-18.04 platform= username= forcePull=false
...
```

## On where the containers run
They run on your local machine, so in *that sense* your repo is 'safe' (i.e. it is not running in someone else's computer).

## On writing to files
If you e.g. ```echo "hello" > output.txt``` in your workflow to create a text file, it will be done *inside the running container*. The container is deleted on exit so you don't see the file.

You could use the ```bind``` flag to ```act```, which binds your working directory to the container rather than copying it. To use the example from above:
```
# act without bind - just shows README in my example project
$ act
...
$ ls
> README.md


# act with bind - shows that a file created inside the action is saved
$ act --bind
...
$ ls
> README.md	output.txt
```

If you are getting different outputs/results to these examples, it would be great if you can post some of the commands you are using and more info about your setup :)

**Reply by Mohammad-Alqadi** [reply]
good day antvirf, i would first thank you for the explanation but i followed what you just mentioned and still i couldn't see the imagine neither the container:
Im using: act --verbose --reuse --bind --secret-file .env -W "./.github_locally" 
![image](https://user-images.githubusercontent.com/104459862/187068116-8935d6ab-3567-4f8d-9994-e91bfc9355c2.png)

The bind flag works fine, what confuses me the most the i have purge all the data of docker and still every time it feels like i caches everything.

**Reply by Antvirf** [reply]
Could you post the same output but also add your final ```act``` run command? Could then see in then output if it is actually pulling or using a cached image to try and answer that part.

Also, I'm not sure if in your example screenshot you had triggered a workflow already? Or whether the screenshot was just a snapshot on your machine without any workflows running?

**Reply by Mohammad-Alqadi** [reply]
The workflow was triggered successfully and the act works perfectly fine like exactly how i want it to be, the only problem im facing is that im not seeing any container nor the image the act uses.
![image](https://user-images.githubusercontent.com/104459862/187075650-2acc9935-b347-46d8-b0ac-7594aff13a48.png)

**Comment by Mohammad-Alqadi** [comment]
Thank you Antvirf, i figure it out.
silly me i just had to add sudo.
